feat: Add incremental `toByteArray` method

ppkarwasz opened the pull request:
This introduces `toByteArray(InputStream input, int size, int bufferSize)`, which reads the stream in chunks of `bufferSize` instead of allocating the full array up front.

By reading incrementally, the method:

* Validates that the stream actually contains `size` bytes before completing the allocation.
* Prevents excessive memory usage if a corrupted or malicious `size` value is provided.
* Offers safer handling for untrusted input compared to the direct-allocation variant.

- [x] I used AI to create the Javadoc.
- [x] Run a successful build using the default [Maven](https://maven.apache.org/) goal with `mvn`; that's `mvn` on the command line by itself.
- [x] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied. This may not always be possible, but it is a best-practice.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Each commit in the pull request should have a meaningful subject line and body. Note that a maintainer may squash commits during the merge process.

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
+     *   If the stream ends prematurely, an {@link EOFException} is thrown.
+     * </p>
+     *
+     * <p>
```
This feels like FUD to me, and I would remove the whole paragraph. I don't think we want to start documenting which methods fall into which defensive category, especially when the actual issue would be in the call site, not this implementation.

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
+     *   This is a convenience overload of {@link #toByteArray(InputStream, int, int)} that accepts a
+     *   {@code long} size parameter. The value is checked to ensure it does not exceed
+     *   {@link Integer#MAX_VALUE} before being safely converted to {@code int}.
+     * </p>
```
This whole paragraph (above) is wrong.

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
-     * @throws IllegalArgumentException if {@code size} is less than zero.
+     * <p>
+     *   This variant allocates the target array immediately and attempts to fill it in one pass.
+     *   It assumes that {@code size} is correct.
```
Let's not say we assume a parameter is correct; after all, we don't say when we assume a parameter incorrect in other cases. That's what exceptions are for, and sometimes the return value.

ppkarwasz replied:
Fixed in https://github.com/apache/commons-io/pull/776/commits/7362d3e34801662b029ea6ef4e005dfb1194c57b

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
+     * <p>
+     *   This variant allocates the target array immediately and attempts to fill it in one pass.
+     *   It assumes that {@code size} is correct.
+     *   If the stream ends prematurely, an {@link EOFException} is thrown.
```
You can remove this since it duplicates the exception documentation.

ppkarwasz replied:
Fixed in https://github.com/apache/commons-io/pull/776/commits/7362d3e34801662b029ea6ef4e005dfb1194c57b

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
-     * @throws IllegalArgumentException if size is less than zero or size is greater than Integer.MAX_VALUE.
-     * @see IOUtils#toByteArray(InputStream, int)
+     * <p>
+     *   This is a convenience overload of {@link #toByteArray(InputStream, int, int)} that accepts a
```
Let's not define convenience versus not; the convenience is only due to the implementation, which feels wrong, we should probably (and separately) fix the implementation to use `Math.toExactInt()`.

ppkarwasz replied:
Since this method throws an unchecked exception if the `long` does not fit into an `int`, most callers have to check the parameter, before calling this. So this methods looks like a convenient composition of a size check and `toByteArray(InputStream, int, int)`.

Fixed in https://github.com/apache/commons-io/pull/776/commits/7362d3e34801662b029ea6ef4e005dfb1194c57b

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2724,62 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+     *
+     * <p>
+     *   This variant validates that the stream actually contains {@code size} bytes.
+     *   It is suitable for untrusted input because it prevents oversized allocations when the provided {@code size}
```
I don't think we should start to document which APIs are designed for trusted versus untrusted input. The documentation should just describe the intended functionality.

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2724,62 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+            throw new IllegalArgumentException("Size must be equal or greater than zero: " + size);
+        }
+        if (bufferSize <= 0) {
+            throw new IllegalArgumentException("Chunk size must be greater than zero: " + bufferSize);
```
The message text and parameter names should match: Either "buffer" or "chunk", not both.

ppkarwasz replied:
Fixed in https://github.com/apache/commons-io/pull/776/commits/c6c79a3dc4fcfcc84a82189f9ae41e3a051a5d35

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2724,62 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+     */
+    public static byte[] toByteArray(final InputStream input, final int size, final int bufferSize) throws IOException {
+        Objects.requireNonNull(input, "input");
+        if (size < 0) {
```
You can remove this check since the worker method does it (in master).

garydgregory reviewed (changes requested):
Hi @ppkarwasz 

I have lots of comments! 😉

ppkarwasz commented on src/main/java/org/apache/commons/io/RandomAccessFiles.java:
```diff
@@ -76,7 +76,11 @@ private static long length(final RandomAccessFile raf) throws IOException {
     public static byte[] read(final RandomAccessFile input, final long position, final int length) throws IOException {
+        Objects.requireNonNull(input, "input");
         input.seek(position);
+        if (length < 0) {
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/d748e99d390eef117d54bf175dcaa569e2b9106d

ppkarwasz commented on src/changes/changes.xml:
```diff
@@ -57,6 +57,7 @@ The <action> type attribute can be add,update,fix,remove.
       <action dev="ggregory" type="add"                due-to="Gary Gregory">Add org.apache.commons.io.output.ProxyOutputStream.writeRepeat(byte[], int, int, long).</action>
       <action dev="ggregory" type="add"                due-to="Gary Gregory">Add org.apache.commons.io.output.ProxyOutputStream.writeRepeat(byte[], long).</action>
       <action dev="ggregory" type="add"                due-to="Gary Gregory">Add org.apache.commons.io.output.ProxyOutputStream.writeRepeat(int, long).</action>
+      <action dev="pkarwasz" type="add"                due-to="Piotr P. Karwasz">Added toByteArray(InputStream, int, int) for safer incremental reading with size validation.</action>
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/33a4fe15a7c05a3b850b6a9872ac6dd3b7c0aaf6

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
     public static byte[] toByteArray(final InputStream input, final int size) throws IOException {
-        return toByteArray(Objects.requireNonNull(input, "input")::read, size);
+        Objects.requireNonNull(input, "input");
+        if (size < 0) {
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/d748e99d390eef117d54bf175dcaa569e2b9106d

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2659,35 +2660,61 @@ public static byte[] toByteArray(final InputStream inputStream) throws IOExcepti
-     * @throws IOException if an I/O error occurs or {@link InputStream} length is smaller than parameter {@code size}.
-     * @throws IllegalArgumentException if {@code size} is less than zero.
+     * <p>
+     *   This variant allocates the target array immediately and attempts to fill it in one pass.
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/7362d3e34801662b029ea6ef4e005dfb1194c57b

ppkarwasz commented:
@garydgregory,

I removed a lot of details from the Javadoc and let only these:

- For `toByteArray(InputStream)`:
  https://github.com/apache/commons-io/blob/c6c79a3dc4fcfcc84a82189f9ae41e3a051a5d35/src/main/java/org/apache/commons/io/IOUtils.java#L2643-L2644
- For `toByteArray(InputStream, int/long)`:
  https://github.com/apache/commons-io/blob/c6c79a3dc4fcfcc84a82189f9ae41e3a051a5d35/src/main/java/org/apache/commons/io/IOUtils.java#L2666-L2667
- For `toByteArray(InputStream, int, int)`:
  https://github.com/apache/commons-io/blob/c6c79a3dc4fcfcc84a82189f9ae41e3a051a5d35/src/main/java/org/apache/commons/io/IOUtils.java#L2708-L2709

I believe that these behaviors will not change and should remain in the method contracts, what do you think?

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2638,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-     * @return the requested byte array.
-     * @throws NullPointerException if the InputStream is {@code null}.
-     * @throws IOException          if an I/O error occurs or reading more than {@link Integer#MAX_VALUE} occurs.
+     * <p>The method accumulates the data in temporary buffers and returns a single array
```
This comment is not needed IMO, the user cannot affect the behavior about it by passing in different values. It's a low-level implementation detail.

ppkarwasz replied:
Removed in https://github.com/apache/commons-io/pull/776/commits/97d37a94de8c935ea8e4961e354a1cd1d5c49b6a

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2638,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-     * <p>
-     * This method buffers the input internally, so there is no need to use a {@link BufferedInputStream}.
-     * </p>
+     * Reads all remaining bytes from the given {@link InputStream} into a new {@code byte[]}.
```
What follows is just about the finer aspects of documenting methods, but it's still nice to talk about IMO:

The text can be simpler and more direct, for example, `java.nio.file.Files.read(InputStream, int)` says "Reads all the bytes from an input stream." After all, the only kinds of bytes in an input stream are the "remaining" kind, and it only makes sense to return a "new" array. So I would say: "Reads all the bytes from an input stream in a byte array".

As an aside, Java seems to have called out the "remaining" concept formally in one API name: "Iterator.forEachRemaining(...)", but that might have been to avoid overriding or shadowing the "forEach(...)".

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2638,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-     * @return byte [] of length {@code size}.
-     * @throws IOException if an I/O error occurs or {@link InputStream} length is smaller than parameter {@code size}.
-     * @throws IllegalArgumentException if {@code size} is less than zero.
+     * <p>The method allocates a single array of the requested size and fills it directly
```
This comment is not needed IMO, the user cannot affect the behavior by passing in different values. It's a low-level implementation detail.

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2638,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-     * @throws IOException              if an I/O error occurs or {@link InputStream} length is less than {@code size}.
-     * @throws IllegalArgumentException if size is less than zero or size is greater than Integer.MAX_VALUE.
-     * @see IOUtils#toByteArray(InputStream, int)
+     * <p>The method allocates a single array of the requested size and fills it directly
```
This comment is not needed IMO, the user cannot affect the behavior by passing in different values. It's a low-level implementation detail.

ppkarwasz replied:
Removed in https://github.com/apache/commons-io/pull/776/commits/97d37a94de8c935ea8e4961e354a1cd1d5c49b6a

garydgregory commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2702,48 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+    /**
+     * Reads exactly {@code size} bytes from the given {@link InputStream} into a new {@code byte[]}.
+     *
+     * <p>The method accumulates the data in temporary buffers of size at most {@code bufferSize}
```
I think it would be nice to explain _why_ this method is useful, instead of _how_ it works. For example, "When reading from an untrusted input stream, you reduce the risk of an OOME by calling this API with...".

garydgregory reviewed (changes requested):
Hi @ppkarwasz ,
I have comments scattered about.
TY!

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2693,52 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+    /**
+     * Reads exactly {@code size} bytes from the given {@link InputStream} into a new {@code byte[]}.
+     *
+     * <p>When reading from an untrusted stream, this variant lowers the risk of
+     * {@link OutOfMemoryError} by allocating data in buffers of up to {@code bufferSize}
+     * bytes rather than in one large array.</p>
+     *
+     * <p>Note, however, that this approach requires additional temporary memory
+     * compared to {@link #toByteArray(InputStream, int)}.</p>
```
If we document `OutOfMemoryError` risk in the Javadoc…

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2693,52 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+        if (bufferSize <= 0) {
+            throw new IllegalArgumentException("Buffer size must be greater than zero: " + bufferSize);
+        }
+        if (size <= bufferSize) {
+            // throws if size < 0
+            return toByteArray(input::read, size);
+        }
```
…should we also extend the "fast path" case to situations where the requested size is already available in the stream?

```suggestion
        // Fast path: use direct allocation if the size fits in one buffer
        // or if at least `size` bytes are already available without blocking.
        if (size <= Math.max(bufferSize, input.available())) {
            // throws if size < 0
            return toByteArray(input::read, size);
        }
```

In this case what would you suggest I add to the Javadoc?

ppkarwasz commented:
Hi @garydgregory,

I’ve refactored this proposal a bit further:

* Extended chunked reading to the legacy `toByteArray(InputStream, int/long)` methods as well.
* Revised the Javadoc to clarify the contract. As you mentioned earlier, we should **not** guide users based on how *trusted* the size parameter is. I’ve also removed explicit references to `OutOfMemoryError`, which can always occur. Instead, the docs now emphasize that memory allocation is **proportional to the number of bytes actually read** (previously it was _de facto_ proportional to the `size` argument).

Open questions:

* **Chunking threshold**: Currently set to 128 KiB, matching CPython. JDK 11+ uses 16 KiB. We could also consider raising our default buffer size (currently 8 KiB, and in some places as low as 1 KiB, which feels outdated).
* **Use of `available()`**: Should we consult `available()` in `toByteArray(InputStream, int)` and `toByteArray(InputStream)` (see https://github.com/apache/commons-io/pull/776#discussion_r2326598113)? The `toByteArray(InputStream, int, int)` method explicitly promises chunks up to `chunkSize`, but the others do not.

garydgregory commented:
> Hi @garydgregory,
> 
> I’ve refactored this proposal a bit further:
> 
> * Extended chunked reading to the legacy `toByteArray(InputStream, int/long)` methods as well.
> * Revised the Javadoc to clarify the contract. As you mentioned earlier, we should **not** guide users based on how _trusted_ the size parameter is. I’ve also removed explicit references to `OutOfMemoryError`, which can always occur. Instead, the docs now emphasize that memory allocation is **proportional to the number of bytes actually read** (previously it was _de facto_ proportional to the `size` argument).
> 
> Open questions:
> 
> * **Chunking threshold**: Currently set to 128 KiB, matching CPython. JDK 11+ uses 16 KiB. We could also consider raising our default buffer size (currently 8 KiB, and in some places as low as 1 KiB, which feels outdated).
> * **Use of `available()`**: Should we consult `available()` in `toByteArray(InputStream, int)` and `toByteArray(InputStream)` (see [feat: Add incremental `toByteArray` method #776 (comment)](https://github.com/apache/commons-io/pull/776#discussion_r2326598113))? The `toByteArray(InputStream, int, int)` method explicitly promises chunks up to `chunkSize`, but the others do not.

CPython is irrelevant IMO. Java's `Files.BUFFER_SIZE` class in Java 21 is 8K, so 8K is consistent as the default.

> JDK 11+ uses 16 KiB.

Where do you see that?

ppkarwasz commented:
> > JDK 11+ uses 16 KiB.
> 
> Where do you see that?

In the [source code of `InputStream`](https://github.com/openjdk/jdk/blob/bea2b029a77e126171d17c3a44baec6d5cafed4a/src/java.base/share/classes/java/io/InputStream.java#L58). This is the value used by `readAllBytes/readNBytes`.

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -221,6 +221,21 @@ public class IOUtils {
+    /**
+     * The maximum size of an array in many Java VMs.
+     */
+    private static final int MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/38a6a2c6b7dcb3cf0646a54ce4a762913f6b7627

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2697,6 +2720,69 @@ public static byte[] toByteArray(final InputStream input, final long size) throw
+
+    /**
+     * Copies up to {@code size} bytes from the given {@link InputStream} into a new {@link UnsynchronizedByteArrayOutputStream}.
+     *
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/29f365be5d485dce7d03f0d0384dbf36fd40cc9d

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2652,65 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-            copy(inputStream, thresholdOutput);
-            return ubaOutput.toByteArray();
+        final UnsynchronizedByteArrayOutputStream output =
+                copyToOutputStream(inputStream, MAX_ARRAY_LENGTH + 1, DEFAULT_CHUNK_SIZE);
```
Fixed in https://github.com/apache/commons-io/pull/776/commits/29f365be5d485dce7d03f0d0384dbf36fd40cc9d

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2652,65 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-    @SuppressWarnings("resource")
     public static byte[] toByteArray(final InputStream input, final int size) throws IOException {
-        return toByteArray(Objects.requireNonNull(input, "input")::read, size);
+        return toByteArray(input, size, DEFAULT_CHUNK_SIZE);
```
I restored the previous behavior in https://github.com/apache/commons-io/pull/776/commits/3afae073c62d55f241f3893612e85ce650b79bbe, but I added a recommendation to use `toByteArray(InputStream, int, int)` for large arrays:

https://github.com/apache/commons-io/blob/4aba097d871332ca1ac537abb5f6880c0bf0ffc3/src/main/java/org/apache/commons/io/IOUtils.java#L2691-L2693

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -221,6 +221,21 @@ public class IOUtils {
+     * for copying streams.
+     * </p>
+     */
+    private static final int DEFAULT_CHUNK_SIZE = 128 * 1024;
```
I switched to 8 KiB in https://github.com/apache/commons-io/pull/776/commits/4aba097d871332ca1ac537abb5f6880c0bf0ffc3

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2652,65 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-            return ubaOutput.toByteArray();
+        final UnsynchronizedByteArrayOutputStream output =
+                copyToOutputStream(inputStream, MAX_ARRAY_LENGTH + 1, DEFAULT_CHUNK_SIZE);
+        if (output.size() > MAX_ARRAY_LENGTH) {
```
Yes, it's possible since `MAX_ARRAY_LENGTH` is `Integer.MAX_VALUE - 8`.

ecki commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2645,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-            copy(inputStream, thresholdOutput);
-            return ubaOutput.toByteArray();
+        final UnsynchronizedByteArrayOutputStream output = copyToOutputStream(inputStream, SOFT_MAX_ARRAY_LENGTH + 1, DEFAULT_BUFFER_SIZE);
+        if (output.size() > SOFT_MAX_ARRAY_LENGTH) {
```
Hm, the UABOS.size() feels a bit overloaded here and why does the write() not throw?

ecki commented on src/test/java/org/apache/commons/io/IOUtilsTest.java:
```diff
@@ -1659,6 +1662,48 @@ void testToByteArray_InputStream_SizeZero() throws Exception {
+        }
+        return Stream.of(
+                // Eager reading
+                Arguments.of(data.clone(), 512, 1024),
```
Those clones and array copies are pretty defensive, but I guess it’s fine here the test only uses small ones

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2652,65 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-     * @return byte [] of length {@code size}.
-     * @throws IOException if an I/O error occurs or {@link InputStream} length is smaller than parameter {@code size}.
-     * @throws IllegalArgumentException if {@code size} is less than zero.
+     * <p>The memory used by this method is <strong>proportional</strong> to the number
```
Yes, I borrowed the **proportional** keyword, which seems to be the magical word to dismiss DoS reports.

Of course `<bytes read> + Integer.MAX_VALUE` is also proportional to the number of bytes read. :wink:

ppkarwasz commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -2637,57 +2645,61 @@ public static BufferedReader toBufferedReader(final Reader reader, final int siz
-    @SuppressWarnings("resource")
     public static byte[] toByteArray(final InputStream input, final int size) throws IOException {
-        return toByteArray(Objects.requireNonNull(input, "input")::read, size);
+        return toByteArray(input::read, size);
```
Nice catch! :100:

Fixed in https://github.com/apache/commons-io/pull/776/commits/15b249da4f4e6bc8238dfce9ae34f4d21f7f88a4

MarkEWaite commented:
As far as I can tell from `git bisect` with builds of commons-io bundled into Jenkins core, this is the pull request that caused a major regression in [Jenkins 2.537](https://www.jenkins.io/changelog/2.537/).  The major regression was:

* https://github.com/jenkinsci/jenkins/issues/16845

Jenkins agents failed to connect with SSH until we reverted Apache Commons IO 2.21.0 and returned to Apache Commons IO 2.20.0.  We released [Jenkins 2.538](https://www.jenkins.io/changelog/2.538/) less than 24 hours after 2.537 due to the severity of the issue.

The stack trace reported in the failure is:

```
[SSH] Starting sftp client.
[SSH] Copying latest remoting.jar...
java.io.IOException: Could not copy remoting.jar into '/home/jenkins/.jenkins-cd-control' on agent
    at PluginClassLoader for ssh-slaves//hudson.plugins.sshslaves.SSHLauncher.copyAgentJar(SSHLauncher.java:738)
    at PluginClassLoader for ssh-slaves//hudson.plugins.sshslaves.SSHLauncher.lambda$launch$0(SSHLauncher.java:462)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalArgumentException: invalid len argument
    at PluginClassLoader for trilead-api//com.trilead.ssh2.SFTPv3Client.read(SFTPv3Client.java:1250)
    at PluginClassLoader for trilead-api//com.trilead.ssh2.jenkins.SFTPClient$SFTPInputStream.read(SFTPClient.java:172)
    at org.apache.commons.io.input.ProxyInputStream.read(ProxyInputStream.java:337)
    at org.apache.commons.io.input.BoundedInputStream.read(BoundedInputStream.java:536)
    at org.apache.commons.io.output.AbstractByteArrayOutputStream.writeImpl(AbstractByteArrayOutputStream.java:405)
    at org.apache.commons.io.output.UnsynchronizedByteArrayOutputStream.write(UnsynchronizedByteArrayOutputStream.java:227)
    at org.apache.commons.io.IOUtils.copyToOutputStream(IOUtils.java:1958)
    at org.apache.commons.io.IOUtils.toByteArray(IOUtils.java:2918)
    at PluginClassLoader for ssh-slaves//hudson.plugins.sshslaves.SSHLauncher.readInputStreamIntoByteArrayAndClose(SSHLauncher.java:796)
    at PluginClassLoader for ssh-slaves//hudson.plugins.sshslaves.SSHLauncher.copyAgentJar(SSHLauncher.java:705)
    ... 5 more
Launch failed - cleaning up connection
[SSH] Connection closed.
```
